pkg/certificate/certificate.go

```diff
@@ -0,0 +1,318 @@
+		extKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
+
+	case UsageServer:
+		// make sure to use the correct key-usage
+		extKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+
+		// And add all the host-ips
+		if hostIps, err = util.GetHostIPs(); err != nil {
+			return nil, nil, nil, humane.Wrap(err, "failed to extract server IPs",
+				"this should never happen",
+				"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+			)
+		}
+
+	default:
+		return nil, nil, nil, humane.New(fmt.Sprintf("invalid certificate usage %s", options.Usage.String()),
+			"this should never happen",
+			"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+		)
+	}
+
+	clientTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(time.Now().UnixNano()),
+		Subject: pkix.Name{
+			CommonName: options.CommonName,
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().Add(365 * 24 * time.Hour),
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage: extKeyUsage,
+		DNSNames:    []string{"localhost", hostname, fmt.Sprintf("%s.local", hostname)},
+		IPAddresses: hostIps,
+	}
+
+	clientKey, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	if err != nil {
+		return nil, nil, nil, humane.Wrap(err, "failed to generate client key",
+			"this should never happen",
+			"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+		)
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, clientTemplate, options.CaCert, &clientKey.PublicKey, options.CaKey)
+	if err != nil {
+		return nil, nil, nil, humane.Wrap(err, "failed to create client certificate",
+			"this should never happen",
+			"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+		)
+	}
+
+	clientKeyBytes, err := x509.MarshalECPrivateKey(clientKey)
+	if err != nil {
+		return nil, nil, nil, humane.Wrap(err, "failed to marshal client private key",
+			"this should never happen",
+			"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+		)
+	}
+
+	switch options.OutputFormat {
+	case FormatPEM:
+		caCertPEM, err := x509.CreateCertificate(rand.Reader, options.CaCert, options.CaCert, &options.CaCert.PublicKey, options.CaKey)
+		if err != nil {
+			return nil, nil, nil, humane.Wrap(err, "failed to marshal client private key",
+				"this should never happen",
+				"please report this as a bug to https://github.com/uptime-industries/compute-blade-agent/issues",
+			)
+		}
+
+		caPEM = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: caCertPEM})
+		certPEM = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
+		keyPEM = pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: clientKeyBytes})
+
+		return caPEM, certPEM, keyPEM, nil
+
+	case FormatDER:
+		return certDER, certDER, clientKeyBytes, nil
```

Comment:
I wonder if we even need to support DER encoding? Wouldn't be PEM sufficient? (might be just me always using PEM)

Reply:
Yeah, honestly, same... I never use anything else than DER encoding. I have no problem giving output format DER the boot :)